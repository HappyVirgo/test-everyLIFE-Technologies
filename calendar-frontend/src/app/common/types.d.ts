import {Event} from 'react-big-calendar';

export interface EltEvent extends Event {
  id?: number;
}
