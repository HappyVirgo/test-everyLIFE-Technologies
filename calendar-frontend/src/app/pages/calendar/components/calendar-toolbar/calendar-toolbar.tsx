import { css } from '@emotion/react';
import { EltEvent } from '../../../../common/types';
import moment from 'moment';
import { random } from 'lodash';
import { Dispatch } from 'react';

interface ICalendarToolbarProps {
  addEvent: (event: Omit<EltEvent, 'id'>) => Promise<void>;
  showIds: boolean;
  setShowIds: Dispatch<boolean>;
  selectedEvent?: EltEvent;
}

export const CalendarToolbar = ({
  addEvent,
  showIds,
  setShowIds,
  selectedEvent,
}: ICalendarToolbarProps) => {
  const createRandomEvent = async () => {
    // Start in the next 48h
    const start = moment().add(random(15, 48 * 60), 'minutes');
    // Last between 30m and 1h 30m
    const end = start.clone().add(random(30, 90), 'minutes');
    const title = `Random event ${random(1, 9999)}`;

    await addEvent({ title, start: start.toDate(), end: end.toDate() });
  };

  const editEvent = (event?: EltEvent) => {
    console.log('todo');
  };

  return (
    <div css={ToolbarStyle}>
      <button data-testid="add-event-btn" onClick={createRandomEvent}>
        Add event
      </button>
      <button
        data-testid="edit-event-btn"
        onClick={() => editEvent(selectedEvent)}
        disabled={!selectedEvent}
      >
        Edit event
      </button>
      <label htmlFor="show-ids-checkbox">
        <input
          id="show-ids-checkbox"
          type="checkbox"
          defaultChecked={showIds}
          onClick={(e) => setShowIds(e.currentTarget.checked)}
        ></input>
        Show ids
      </label>
    </div>
  );
};

const ToolbarStyle = css`
  width: 100%;
  text-align: center;
  padding: 8px 0;
  margin-bottom: 8px;
  border-bottom: 1px solid;
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 10px;

  button {
    color: #373a3c;
    display: inline-block;
    margin: 0;
    text-align: center;
    vertical-align: middle;
    background: none;
    border: 1px solid #ccc;
    padding: 0.375rem 1rem;
    border-radius: 4px;
    line-height: normal;
    white-space: nowrap;

    &:hover {
      color: #373a3c;
      background-color: rgb(229.5, 229.5, 229.5);
      border-color: rgb(173.4, 173.4, 173.4);
      cursor: pointer;
    }

    &:active {
      box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.125);
      background-color: rgb(229.5, 229.5, 229.5);
      border-color: rgb(173.4, 173.4, 173.4);
    }

    &:disabled {
      background-color: #ccc;
      color: #fff;
      pointer-events: none;
    }
  }
`;
