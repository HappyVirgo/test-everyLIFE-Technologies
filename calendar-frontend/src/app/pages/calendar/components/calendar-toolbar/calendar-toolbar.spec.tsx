import { render, screen } from '@testing-library/react';
import { CalendarToolbar } from './calendar-toolbar';
import { EltEvent } from '../../../../common/types';
import { Dispatch } from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

describe('CalendarToolbarComponent', () => {
  let addEvent: (event: Omit<EltEvent, 'id'>) => Promise<void>;
  let setShowIds: Dispatch<boolean>;
  const mockEvent: EltEvent = {
    id: 100,
    title: 'Mock event',
    start: new Date(),
    end: new Date(),
  };

  beforeEach(() => {
    addEvent = jest.fn();
    setShowIds = jest.fn();
  });

  it('renders correctly', () => {
    const { container } = render(
      <CalendarToolbar
        addEvent={addEvent}
        showIds={false}
        setShowIds={setShowIds}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  describe('Add event button', () => {
    it('should add a random event', async () => {
      render(
        <CalendarToolbar
          addEvent={addEvent}
          showIds={false}
          setShowIds={setShowIds}
        />,
      );

      const btn = screen.getByTestId('add-event-btn');
      userEvent.click(btn);

      expect(addEvent).toHaveBeenCalledWith({
        start: expect.any(Date),
        end: expect.any(Date),
        title: expect.stringMatching(/Random event \d+/),
      });
    });
  });

  describe('Edit event button', () => {
    it('should only be disabled if there is no selected event', async () => {
      render(
        <CalendarToolbar
          addEvent={addEvent}
          showIds={false}
          setShowIds={setShowIds}
        />,
      );

      const btn = screen.getByTestId('edit-event-btn');
      expect(btn).toBeDisabled();
    });

    it('should only be disabled if there is a selected event', async () => {
      render(
        <CalendarToolbar
          addEvent={addEvent}
          showIds={false}
          setShowIds={setShowIds}
          selectedEvent={mockEvent}
        />,
      );

      const btn = screen.getByTestId('edit-event-btn');
      expect(btn).toBeEnabled();
    });
  });

  describe('Show ids checkbox', () => {
    it('should toggle ids being shown', () => {
      render(
        <CalendarToolbar
          addEvent={addEvent}
          showIds={false}
          setShowIds={setShowIds}
        />,
      );

      const checkbox = screen.getByLabelText('Show ids');
      expect(checkbox).not.toBeChecked();

      // Check
      userEvent.click(checkbox);
      expect(checkbox).toBeChecked();
      expect(setShowIds).toHaveBeenCalledWith(true);

      // Uncheck
      userEvent.click(checkbox);
      expect(checkbox).not.toBeChecked();
      expect(setShowIds).toHaveBeenCalledWith(false);
    });
  });
});
