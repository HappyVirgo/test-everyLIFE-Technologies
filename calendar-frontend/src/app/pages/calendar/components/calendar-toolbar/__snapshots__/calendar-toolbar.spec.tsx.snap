// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarToolbarComponent renders correctly 1`] = `
<div>
  <div
    css="[object Object]"
  >
    <button
      data-testid="add-event-btn"
    >
      Add event
    </button>
    <button
      data-testid="edit-event-btn"
      disabled=""
    >
      Edit event
    </button>
    <label
      for="show-ids-checkbox"
    >
      <input
        id="show-ids-checkbox"
        type="checkbox"
      />
      Show ids
    </label>
  </div>
</div>
`;
